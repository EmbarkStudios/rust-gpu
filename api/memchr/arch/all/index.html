<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Contains architecture independent routines."><title>memchr::arch::all - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../../static.files/rustdoc-fa3bb1812debf86c.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="memchr" data-themes="" data-resource-suffix="" data-rustdoc-version="1.74.0-nightly (8ce4540bd 2023-09-29)" data-channel="nightly" data-search-js="search-8be46b629f5f14a8.js" data-settings-js="settings-74424d7eec62a23e.js" ><script src="../../../static.files/storage-fec3eaa3851e447d.js"></script><script defer src="../../../static.files/main-c5bd66d33317d69f.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-5d8b3c7633ad77ba.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../../../memchr/index.html"><img class="rust-logo" src="../../../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a></nav><nav class="sidebar"><a class="logo-container" href="../../../memchr/index.html"><img class="rust-logo" src="../../../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location"><a href="#">Module all</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#modules">Modules</a></li><li><a href="#functions">Functions</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../../../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../../index.html">memchr</a>::<wbr><a href="../index.html">arch</a>::<wbr><a class="mod" href="#">all</a><button id="copy-path" title="Copy item path to clipboard"><img src="../../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../../../src/memchr/arch/all/mod.rs.html#1-234">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Contains architecture independent routines.</p>
<p>These routines are often used as a “fallback” implementation when the more
specialized architecture dependent routines are unavailable.</p>
</div></details><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="memchr/index.html" title="mod memchr::arch::all::memchr">memchr</a></div><div class="desc docblock-short">Provides architecture independent implementations of <code>memchr</code> and friends.</div></li><li><div class="item-name"><a class="mod" href="packedpair/index.html" title="mod memchr::arch::all::packedpair">packedpair</a></div><div class="desc docblock-short">Provides an architecture independent implementation of the “packed pair”
algorithm.</div></li><li><div class="item-name"><a class="mod" href="rabinkarp/index.html" title="mod memchr::arch::all::rabinkarp">rabinkarp</a></div><div class="desc docblock-short">An implementation of the <a href="https://en.wikipedia.org/wiki/Rabin%E2%80%93Karp_algorithm">Rabin-Karp substring search algorithm</a>.</div></li><li><div class="item-name"><a class="mod" href="shiftor/index.html" title="mod memchr::arch::all::shiftor">shiftor</a></div><div class="desc docblock-short">An implementation of the <a href="https://en.wikipedia.org/wiki/Bitap_algorithm">Shift-Or substring search algorithm</a>.</div></li><li><div class="item-name"><a class="mod" href="twoway/index.html" title="mod memchr::arch::all::twoway">twoway</a></div><div class="desc docblock-short">An implementation of the <a href="https://en.wikipedia.org/wiki/Two-way_string-matching_algorithm">Two-Way substring search algorithm</a>.</div></li></ul><h2 id="functions" class="small-section-header"><a href="#functions">Functions</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.is_equal.html" title="fn memchr::arch::all::is_equal">is_equal</a></div><div class="desc docblock-short">Compare corresponding bytes in <code>x</code> and <code>y</code> for equality.</div></li><li><div class="item-name"><a class="fn" href="fn.is_equal_raw.html" title="fn memchr::arch::all::is_equal_raw">is_equal_raw</a><sup title="unsafe function">⚠</sup></div><div class="desc docblock-short">Compare <code>n</code> bytes at the given pointers for equality.</div></li><li><div class="item-name"><a class="fn" href="fn.is_prefix.html" title="fn memchr::arch::all::is_prefix">is_prefix</a></div><div class="desc docblock-short">Returns true if and only if <code>needle</code> is a prefix of <code>haystack</code>.</div></li><li><div class="item-name"><a class="fn" href="fn.is_suffix.html" title="fn memchr::arch::all::is_suffix">is_suffix</a></div><div class="desc docblock-short">Returns true if and only if <code>needle</code> is a suffix of <code>haystack</code>.</div></li></ul></section></div></main></body></html>