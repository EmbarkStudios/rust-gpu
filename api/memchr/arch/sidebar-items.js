window.SIDEBAR_ITEMS = {"mod":["aarch64","all"]};