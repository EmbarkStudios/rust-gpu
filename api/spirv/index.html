<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="The SPIR-V header."><meta name="keywords" content="rust, rustlang, rust-lang, spirv"><title>spirv - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script src="../crates.js"></script><script defer src="../main.js"></script>
    <noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../spirv/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../spirv/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"><a href="#">Crate spirv</a></h2><div class="sidebar-elems"><div class="block"><ul><li class="version">Version 0.2.0+1.5.4</li><li><a id="all-types" href="all.html">All Items</a></li></ul></div><section><div class="block"><ul><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#constants">Constants</a></li><li><a href="#types">Type Definitions</a></li></ul></div></section><div id="sidebar-vars" data-name="spirv" data-ty="mod" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../spirv/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><button type="button" id="help-button" title="help">?</button><div id="settings-menu" tabindex="-1">
                                <a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div>
                        </div></form></nav></div><section id="main-content" class="content"><div class="main-heading">
    <h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">spirv</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/spirv/lib.rs.html#1-14">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>The SPIR-V header.</p>
<p>This crate contains Rust definitions of all SPIR-V structs, enums,
and constants.</p>
<p>The version of this crate is the version of SPIR-V it contains.</p>
</div></details><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.FPFastMathMode.html" title="spirv::FPFastMathMode struct">FPFastMathMode</a></div><div class="item-right docblock-short"><p>SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_fast_math_mode_a_fp_fast_math_mode">FPFastMathMode</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.FragmentShadingRate.html" title="spirv::FragmentShadingRate struct">FragmentShadingRate</a></div><div class="item-right docblock-short"><p>SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fragment_shading_rate_a_fragment_shading_rate">FragmentShadingRate</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.FunctionControl.html" title="spirv::FunctionControl struct">FunctionControl</a></div><div class="item-right docblock-short"><p>SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_function_control_a_function_control">FunctionControl</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.ImageOperands.html" title="spirv::ImageOperands struct">ImageOperands</a></div><div class="item-right docblock-short"><p>SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_operands_a_image_operands">ImageOperands</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.KernelProfilingInfo.html" title="spirv::KernelProfilingInfo struct">KernelProfilingInfo</a></div><div class="item-right docblock-short"><p>SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_kernel_profiling_info_a_kernel_profiling_info">KernelProfilingInfo</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.LoopControl.html" title="spirv::LoopControl struct">LoopControl</a></div><div class="item-right docblock-short"><p>SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_loop_control_a_loop_control">LoopControl</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.MemoryAccess.html" title="spirv::MemoryAccess struct">MemoryAccess</a></div><div class="item-right docblock-short"><p>SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_access_a_memory_access">MemoryAccess</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.MemorySemantics.html" title="spirv::MemorySemantics struct">MemorySemantics</a></div><div class="item-right docblock-short"><p>SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_semantics_a_memory_semantics">MemorySemantics</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.RayFlags.html" title="spirv::RayFlags struct">RayFlags</a></div><div class="item-right docblock-short"><p>SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_flags_a_ray_flags">RayFlags</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.SelectionControl.html" title="spirv::SelectionControl struct">SelectionControl</a></div><div class="item-right docblock-short"><p>SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_selection_control_a_selection_control">SelectionControl</a></p>
</div></div></div><h2 id="enums" class="small-section-header"><a href="#enums">Enums</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.AccessQualifier.html" title="spirv::AccessQualifier enum">AccessQualifier</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_access_qualifier_a_access_qualifier">AccessQualifier</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.AddressingModel.html" title="spirv::AddressingModel enum">AddressingModel</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_addressing_model_a_addressing_model">AddressingModel</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.BuiltIn.html" title="spirv::BuiltIn enum">BuiltIn</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_built_in_a_built_in">BuiltIn</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.CLOp.html" title="spirv::CLOp enum">CLOp</a></div><div class="item-right docblock-short"><p><a href="https://www.khronos.org/registry/spir-v/specs/unified1/OpenCL.ExtendedInstructionSet.100.html">OpenCL.std</a> extended instruction opcode</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.Capability.html" title="spirv::Capability enum">Capability</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_capability_a_capability">Capability</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.Decoration.html" title="spirv::Decoration enum">Decoration</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_decoration_a_decoration">Decoration</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.Dim.html" title="spirv::Dim enum">Dim</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_dim_a_dim">Dim</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.ExecutionMode.html" title="spirv::ExecutionMode enum">ExecutionMode</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_execution_mode_a_execution_mode">ExecutionMode</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.ExecutionModel.html" title="spirv::ExecutionModel enum">ExecutionModel</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_execution_model_a_execution_model">ExecutionModel</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.FPRoundingMode.html" title="spirv::FPRoundingMode enum">FPRoundingMode</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_rounding_mode_a_fp_rounding_mode">FPRoundingMode</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.FunctionParameterAttribute.html" title="spirv::FunctionParameterAttribute enum">FunctionParameterAttribute</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_function_parameter_attribute_a_function_parameter_attribute">FunctionParameterAttribute</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.GLOp.html" title="spirv::GLOp enum">GLOp</a></div><div class="item-right docblock-short"><p><a href="https://www.khronos.org/registry/spir-v/specs/unified1/GLSL.std.450.html">GLSL.std.450</a> extended instruction opcode</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.GroupOperation.html" title="spirv::GroupOperation enum">GroupOperation</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_group_operation_a_group_operation">GroupOperation</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.ImageChannelDataType.html" title="spirv::ImageChannelDataType enum">ImageChannelDataType</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_channel_data_type_a_image_channel_data_type">ImageChannelDataType</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.ImageChannelOrder.html" title="spirv::ImageChannelOrder enum">ImageChannelOrder</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_channel_order_a_image_channel_order">ImageChannelOrder</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.ImageFormat.html" title="spirv::ImageFormat enum">ImageFormat</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_format_a_image_format">ImageFormat</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.KernelEnqueueFlags.html" title="spirv::KernelEnqueueFlags enum">KernelEnqueueFlags</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_kernel_enqueue_flags_a_kernel_enqueue_flags">KernelEnqueueFlags</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.LinkageType.html" title="spirv::LinkageType enum">LinkageType</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_linkage_type_a_linkage_type">LinkageType</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.MemoryModel.html" title="spirv::MemoryModel enum">MemoryModel</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_model_a_memory_model">MemoryModel</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.Op.html" title="spirv::Op enum">Op</a></div><div class="item-right docblock-short"><p>SPIR-V <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_instructions_a_instructions">instructions</a> opcodes</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.RayQueryCandidateIntersectionType.html" title="spirv::RayQueryCandidateIntersectionType enum">RayQueryCandidateIntersectionType</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_candidate_intersection_type_a_ray_query_candidate_intersection_type">RayQueryCandidateIntersectionType</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.RayQueryCommittedIntersectionType.html" title="spirv::RayQueryCommittedIntersectionType enum">RayQueryCommittedIntersectionType</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_committed_intersection_type_a_ray_query_committed_intersection_type">RayQueryCommittedIntersectionType</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.RayQueryIntersection.html" title="spirv::RayQueryIntersection enum">RayQueryIntersection</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_intersection_a_ray_query_intersection">RayQueryIntersection</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.SamplerAddressingMode.html" title="spirv::SamplerAddressingMode enum">SamplerAddressingMode</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_sampler_addressing_mode_a_sampler_addressing_mode">SamplerAddressingMode</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.SamplerFilterMode.html" title="spirv::SamplerFilterMode enum">SamplerFilterMode</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_sampler_filter_mode_a_sampler_filter_mode">SamplerFilterMode</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.Scope.html" title="spirv::Scope enum">Scope</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_scope_a_scope">Scope</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.SourceLanguage.html" title="spirv::SourceLanguage enum">SourceLanguage</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_source_language_a_source_language">SourceLanguage</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.StorageClass.html" title="spirv::StorageClass enum">StorageClass</a></div><div class="item-right docblock-short"><p>/// SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_storage_class_a_storage_class">StorageClass</a></p>
</div></div></div><h2 id="constants" class="small-section-header"><a href="#constants">Constants</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="constant" href="constant.MAGIC_NUMBER.html" title="spirv::MAGIC_NUMBER constant">MAGIC_NUMBER</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="constant" href="constant.MAJOR_VERSION.html" title="spirv::MAJOR_VERSION constant">MAJOR_VERSION</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="constant" href="constant.MINOR_VERSION.html" title="spirv::MINOR_VERSION constant">MINOR_VERSION</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="constant" href="constant.REVISION.html" title="spirv::REVISION constant">REVISION</a></div><div class="item-right docblock-short"></div></div></div><h2 id="types" class="small-section-header"><a href="#types">Type Definitions</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="type" href="type.Word.html" title="spirv::Word type">Word</a></div><div class="item-right docblock-short"></div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="spirv" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.63.0-nightly (546c826f0 2022-06-12)" ></div>
</body></html>