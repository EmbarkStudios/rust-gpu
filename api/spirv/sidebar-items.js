window.SIDEBAR_ITEMS = {"constant":["MAGIC_NUMBER","MAJOR_VERSION","MINOR_VERSION","REVISION"],"enum":["AccessQualifier","AddressingModel","BuiltIn","CLOp","Capability","Decoration","Dim","ExecutionMode","ExecutionModel","FPRoundingMode","FunctionParameterAttribute","GLOp","GroupOperation","ImageChannelDataType","ImageChannelOrder","ImageFormat","KernelEnqueueFlags","LinkageType","MemoryModel","Op","RayQueryCandidateIntersectionType","RayQueryCommittedIntersectionType","RayQueryIntersection","SamplerAddressingMode","SamplerFilterMode","Scope","SourceLanguage","StorageClass"],"struct":["FPFastMathMode","FragmentShadingRate","FunctionControl","ImageOperands","KernelProfilingInfo","LoopControl","MemoryAccess","MemorySemantics","RayFlags","SelectionControl"],"type":["Word"]};