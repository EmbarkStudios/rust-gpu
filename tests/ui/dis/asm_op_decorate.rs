// build-pass
// compile-flags: -C target-feature=+RuntimeDescriptorArray,+ext:SPV_EXT_descriptor_indexing
// compile-flags: -C llvm-args=--disassemble-globals
// normalize-stderr-test "OpCapability VulkanMemoryModel\n" -> ""
// normalize-stderr-test "OpExtension .SPV_KHR_vulkan_memory_model.\n" -> ""
// normalize-stderr-test "OpMemoryModel Logical Vulkan" -> "OpMemoryModel Logical Simple"

// FIXME(eddyb) this should use revisions to track both the `vulkan1.2` output
// and the pre-`vulkan1.2` output, but per-revisions `{only,ignore}-*` directives
// are not supported in `compiletest-rs`.
// ignore-vulkan1.2

use core::arch::asm;
use spirv_std::spirv;

fn add_decorate() {
    unsafe {
        let offset = 1u32;
        asm!(
            "OpDecorate %image_2d_var DescriptorSet 0",
            "OpDecorate %image_2d_var Binding 0",
            "%uint                  = OpTypeInt 32 0",
            "%float                 = OpTypeFloat 32",
            "%uint_0                = OpConstant %uint 0",
            "%image_2d              = OpTypeImage %float Dim2D 0 0 0 1 Unknown",
            "%sampled_image_2d      = OpTypeSampledImage %image_2d",
            "%image_array           = OpTypeRuntimeArray %sampled_image_2d",
            // NOTE(eddyb) `Generic` is used here because it's the placeholder
            // for storage class inference - both of the two `OpTypePointer`
            // types below should end up inferring to `UniformConstant`.
            "%ptr_image_array       = OpTypePointer Generic %image_array",
            "%image_2d_var          = OpVariable %ptr_image_array UniformConstant",
            "%ptr_sampled_image_2d  = OpTypePointer Generic %sampled_image_2d",
            "", // ^^ type preamble
            "%offset                = OpLoad _ {0}",
            "%24                    = OpAccessChain %ptr_sampled_image_2d %image_2d_var %offset",
            "%25                    = OpLoad %sampled_image_2d %24",
            in(reg) &offset,
        );
    }
}
#[spirv(fragment)]
pub fn main() {
    add_decorate();
}
