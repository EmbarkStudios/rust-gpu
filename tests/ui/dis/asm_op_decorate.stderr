OpCapability RuntimeDescriptorArray
OpCapability Shader
OpExtension "SPV_EXT_descriptor_indexing"
OpMemoryModel Logical Simple
OpEntryPoint Fragment %1 "main"
OpExecutionMode %1 OriginUpperLeft
%2 = OpString "$DIR/asm_op_decorate.rs"
OpName %3 "asm_op_decorate::add_decorate"
OpName %4 "asm_op_decorate::main"
OpDecorate %5 ArrayStride 4
OpDecorate %6 DescriptorSet 0
OpDecorate %6 Binding 0
%7 = OpTypeVoid
%8 = OpTypeFunction %7
%9 = OpTypeInt 32 0
%10 = OpConstant  %9  1
%11 = OpTypeFloat 32
%12 = OpTypeImage %11 2D 0 0 0 1 Unknown
%13 = OpTypeSampledImage %12
%5 = OpTypeRuntimeArray %13
%14 = OpTypePointer UniformConstant %5
%6 = OpVariable  %14  UniformConstant
%15 = OpTypePointer UniformConstant %13
